### ISSUE TITLE
Add support for syncing only on InsertLeave

### ISSUE THREAD

**Issue by dkasak** [natebosch/vim-lsc#169]
This is a feature request for a switch which would mean "only sync on `InsertLeave`", as per [this reddit comment](https://old.reddit.com/r/vim/comments/9zo98c/what_languageserver_client_does_everyone_use/eaf2pqi/).

I'm not the OP there, but I'd also love this feature as I'm currently considering switching away from LanguageClient-neovim due to the issues described in that thread.

The ergonomically ideal way of handling diagnostics would be to sync periodically (and not only on `InsertLeave`, just like now), but only update the diagnostics instantly when diagnostics *disappear*, delaying *appearing* diagnostics until a future `InsertLeave` (or a timer expiring, for people that prefer that). I realize this is a much more involved feature, but I wanted to mention it for documentation purposes, in case anyone gets a bout of inspiration to implement it. This is discussed in autozimu/LanguageClient-neovim#603 (e.g. [this comment](https://github.com/autozimu/LanguageClient-neovim/issues/603#issuecomment-469076696)).

**Comment by natebosch** [comment]
> only update the diagnostics instantly when diagnostics disappear, delaying appearing diagnostics until a future `InsertLeave`

Seems interesting - I'd be inclined to aim for that as an option over delaying the sync process. I'd have to look to see how feasible it might be.

I assume a global option would be best for this?

**Comment by natebosch** [comment]
The difficulty here became immediately obvious when I started digging in - I can't thing of a reasonable way to "diff" the diagnostics from what they were previously. More specifically I can't distinguish between a diagnostic that was _moved_ from one being removed and a new one added. This will come up anytime you add a new line in insert mode - all the diagnostics below that line will look new and so they'd disappear.

I wonder if instead if might be better to exclude diagnostic within some range of the cursor. How often is the distraction caused by diagnostic highlighting on the current line compared to other lines?

**Comment by andymass** [comment]
Admittedly I am not following all the linked conversations, but is the "diff" issue you mentioned just that you can't keep track of diagnostics when the text is being edited?  The "text properties" feature was designed to allow placing line/col locations which are automatically updated during editing.  Though it's pretty new.

**Comment by natebosch** [comment]
@dkasak - could you try the branch in https://github.com/natebosch/vim-lsc/pull/173 and let me know your opinion?

It has some unintended side effects of clearing otherwise unaffected diagnostic highlighting for the current line as you type, but that might not be a big deal in practice? Once you leave insert mode, or are typing on a new line and there is an update to the diagnostics, they'll appear.

If you like the behavior of that branch I can either make it an option, or maybe even the default.

**Comment by natebosch** [comment]
> is the "diff" issue you mentioned just that you can't keep track of diagnostics when the text is being edited?

Yeah, if I'm typing and I insert a new line all the diagnostics that _were_ online 5 might now be on line 6, but I can't tell when I get the update that it was a moved diagnostic as opposed to a fresh one. Similar problem if things shift by characters instead of lines.

> The "text properties" feature was designed to allow placing line/col locations which are automatically updated during editing. Though it's pretty new.

I'm not aware any feature called "text properties". I see a `DiagnosticRelatedInformation` which is new from the last time I looked at the spec, but I don't see how that would relate to tracking the identity of a diagnostic across updates.

**Comment by andymass** [comment]
Sorry, text properties is a feature of vim not lsp.  Basically like marks except they track columns and can use highlighting.  I'm thinking you'd add the received diagnostics as text properties, and then query their locations from vim when you need to identify them again.  The other benefit is highlighting moves with the text.  Actually, if you are just interested in line numbers, I think you could already do this with signs.

**Comment by natebosch** [comment]
> text properties is a feature of vim 

Ah cool - I'll check this out. The new concern would likely be that I'd need to take snapshots of the text property locations at the time I sync, but that may be feasible to do. I'll experiment a bit and see what I can can come up with.

Thanks for pointing this out!
